### PR TITLE
BoxRenderObjClass

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/w3d/renderer/boxrobj.h [hunk 1]
```diff
@@ -20,6 +20,9 @@
 #include "vector3.h"
 #include "w3d_file.h"
 #include "w3dmpo.h"
+#include <cstring>
+
+using std::strcpy;
```
Comment (duncanspumpkin) [inline src/w3d/renderer/boxrobj.h]: never put using statements in headers you are just negating the whole reason a namespace exists.

File: src/w3d/renderer/boxrobj.cpp [hunk 2]
```diff
@@ -13,36 +13,499 @@
...
+BoxRenderObjClass::BoxRenderObjClass(const BoxRenderObjClass &src)
+{
+    *this = src;
+}
+
+BoxRenderObjClass::BoxRenderObjClass(const W3dBoxStruct &src) : m_opacity(0.25)
+{
+    Set_Name(src.Name);
+    Convert_Color(src.Color, &m_color);
+    m_objSpaceCenter.X = src.Center.x;
```
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.cpp]: this used Convert_Vector

File: src/w3d/renderer/boxrobj.cpp [hunk 3]
```diff
@@ -13,36 +13,499 @@
...
+    m_objSpaceCenter.Y = src.Center.y;
+    m_objSpaceCenter.Z = src.Center.z;
+    m_objSpaceExtent.X = src.Extent.x;
+    m_objSpaceExtent.Y = src.Extent.y;
+    m_objSpaceExtent.Z = src.Extent.z;
+}
```
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.cpp:130]: here it sets a bitflags and opacity to 0.25

File: src/w3d/renderer/boxrobj.h [hunk 4]
```diff
@@ -41,10 +46,10 @@ class BoxRenderObjClass : public RenderObjClass
 
     BoxRenderObjClass &operator=(const BoxRenderObjClass &that);
 
-    virtual ~BoxRenderObjClass() override;
-    virtual const char *Get_Name() const override;
-    virtual void Set_Name(const char *name) override;
-    virtual int Get_Num_Polys() const override;
+    virtual ~BoxRenderObjClass() override {}
+    virtual const char *Get_Name() const override { return m_name; }
+    virtual void Set_Name(const char *name) override { strcpy(m_name, name); }
```
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.h]: move to cpp, add asserts

File: src/w3d/renderer/boxrobj.cpp [hunk 5]
```diff
@@ -13,36 +13,504 @@
...
+        static Vector3 verts[8];
+        for (int i = 0; i < 8; i++) {
+            verts[i].X = g_boxVerts[i].X * extent.X + center.X;
+            verts[i].Y = g_boxVerts[i].Y * extent.Y + center.Y;
+            verts[i].Z = g_boxVerts[i].Z * extent.Z + center.Z;
+        }
+
+        unsigned int color = DX8Wrapper::Convert_Color(m_color, m_opacity);
+
+        DynamicVBAccessClass vb_access(3, D3DFVF_XYZ | D3DFVF_DIFFUSE, 8);
```
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.cpp]: D3DFVF_XYZ|D3DFVF_NORMAL|D3DFVF_DIFFUSE|D3DFVF_TEX2

File: src/w3d/renderer/boxrobj.cpp [hunk 6]
```diff
@@ -13,36 +13,504 @@
...
+                v[i].z = verts[i].Z;
+
+                v[i].nx = g_boxVertexNormals[i].X;
+                v[i].ny = g_boxVertexNormals[i].Y;
+                v[i].nz = g_boxVertexNormals[i].Z;
+                v[i].diffuse = color;
+            }
+        }
+
+        DynamicIBAccessClass ib_access(3, 36);
```
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.cpp]: 2 not 3

File: src/w3d/renderer/boxrobj.cpp (hunk 5 above)
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.cpp]: 2 not 3

File: src/w3d/renderer/boxrobj.cpp [hunk 7]
```diff
@@ -13,36 +13,504 @@
...
+
+            for (int i = 0; i < 12; i++) {
+                indexes[i] = g_boxFaces[i].I;
+                indexes[i + 1] = g_boxFaces[i].J;
+                indexes[i + 2] = g_boxFaces[i].K;
+            }
+        }
+
+        DX8Wrapper::Set_Shader(g_boxShader);
+        DX8Wrapper::Set_Material(g_boxMaterial);
```
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.cpp:197]: should be first

File: src/w3d/renderer/boxrobj.cpp [hunk 8]
```diff
@@ -13,36 +13,504 @@
...
+        DX8Wrapper::Set_Texture(0, nullptr);
+        DX8Wrapper::Set_Index_Buffer(ib_access, 0);
+        DX8Wrapper::Set_Vertex_Buffer(vb_access);
+        DX8Wrapper::Draw_Triangles(2, 0, 12, 0, 8);
```
Comment (tomsons26) [inline src/w3d/renderer/boxrobj.cpp:204]: theres a virtual call before this

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**